### DISCUSSION TITLE
There is a problem with the assignment of the sp register, when ucontext == NULL

### DISCUSSION THREAD

**Discussion by byte0100** [async-profiler/async-profiler#1041] [Q&A] [answered]
According to the [doc of gcc](https://gcc.gnu.org/onlinedocs/gcc/Return-Address.html#index-_005f_005fbuiltin_005fframe_005faddress), __builtin_frame_address(0) return the frame address of the current function. 

**NOTE** the rsp of the caller is not equal the frame address of the current functon, which is that (rsp - 16)== fp(callee)，in x86-64 arch. 

When cfa_reg == DW_REG_SP , the sp of caller would  be miscalculated because of the wrong sp, which come from __builtin_frame_address(0).

Therefore, the pc also becomes incorrect, resulting in unknown stack frames. Howerver, the subsequent backstack will use FrameDesc::default_frame, that is, cfa_reg == DW_REG_FP, and is not affected by the incorrect sp.

**Comment by apangin** [comment]
Thank you for your interest in async-profiler.
Given that the topic is raised in Q&A section, can you please clarify what the question is here? What kind of answer are you looking for?

To me, this seems to be more like an issue. If this is the case, we encourage you to submit a [bug report](https://github.com/async-profiler/async-profiler/issues/new?assignees=&labels=bug&projects=&template=bug-report.yml&title=%28short+issue+description%29) filling in all relevant fields. In particular, an example that demonstrates the problem would be very much helpful.

**Comment by byte0100** [comment]
Thanks for your reply.

**My question is that the sp of ucontext should be “__builtin_frame_address(0) + LINKED_FRAME_SIZE” , in x86. Is that right?**

int StackWalker::walkDwarf(void* ucontext, const void** callchain, int max_depth, StackContext* java_ctx) {
    const void* pc;
    uintptr_t fp;
    uintptr_t sp;
    uintptr_t bottom = (uintptr_t)&sp + MAX_WALK_SIZE;

    StackFrame frame(ucontext);
    if (ucontext == NULL) {
        pc = __builtin_return_address(0);
        fp = (uintptr_t)__builtin_frame_address(1);
        sp = (uintptr_t)__builtin_frame_address(0);  // <--- this line!!!!
    } else {
        pc = (const void*)frame.pc();
        fp = frame.fp();
        sp = frame.sp();
    }

**Reply by apangin** [reply, marked as answer]
Thank you for clarification. You are right.
There is no such problem in the original async-profiler, since `libasyncProfiler.so` is compiled with frame pointers enabled.
However, when I recompiled it with `-fomit-frame-pointer`, I could observe the mentioned issue in `alloc` mode with `cstack=dwarf`.
I pushed the fix as proposed.
